### PR TITLE
fix(ai): accept assistant tool-call messages that omit content

### LINKED ISSUE #1748: Model Gateway rejects OpenAI-compatible assistant messages that omit content
### Describe the bug

The Model Gateway is branded OpenAI-compatible, but its chat-completion request validation rejects a valid OpenAI message shape: an **assistant message that carries `tool_calls` and omits `content` entirely**.

OpenAI's Chat Completions spec treats assistant `content` as **optional when `tool_calls` is present** (the OpenAI SDK types `content` as optional + nullable on assistant messages). Clients following that convention — hand-rolled JSON, Go with `omitempty`, trimmed payloads — send the assistant turn without a `content` key at all. Those requests work against `api.openai.com` but get a `400 INVALID_INPUT` here.

### Root cause

In [`packages/shared-schemas/src/ai-api.schema.ts`](https://github.com/InsForge/InsForge/blob/main/packages/shared-schemas/src/ai-api.schema.ts), `chatMessageSchema.content` is:

```ts
content: z.union([z.string(), z.array(contentSchema)]).nullable(),
```

`.nullable()` allows `null` but **not** an omitted key, so the `content` property must be physically present. The route validates every request with `chatCompletionRequestSchema.safeParse(req.body)` (`backend/src/api/routes/ai/index.routes.ts`) and throws `INVALID_INPUT` on failure.

The downstream `formatMessages` already tolerates a missing assistant content (`content: ... ?? null` in `chat-completion.service.ts`), so **only the schema is over-strict**.

### Steps to reproduce

`POST /api/ai/chat/completion` with a multi-turn tool conversation where the assistant turn omits `content`:

```json
{
  "model": "openai/gpt-4",
  "messages": [
    { "role": "user", "content": "What is the weather in Tokyo?" },
    { "role": "assistant", "tool_calls": [
      { "id": "call_1", "type": "function",
        "function": { "name": "get_weather", "arguments": "{\"city\":\"Tokyo\"}" } }
    ] },
    { "role": "tool", "content": "{\"temp\":\"22°C\"}", "tool_call_id": "call_1" }
  ]
}
```

- **Actual:** `400` — `Validation error: Required` at `messages[1].content`.
- **Expected:** accepted (the same request with `"content": null` added to the assistant message already succeeds).

### Proposed fix

Relax `content` to `.nullish()` (nullable **and** optional). This matches OpenAI's contract and stays consistent with the schema's existing permissiveness (it already accepts `content: null` for any role); `formatMessages` already coerces a missing assistant content to null, so no downstream change is required.

I'd like to work on this — PR incoming.

### PR DESCRIPTION
**Pull request by hassaanch23**
## Summary

The Model Gateway is OpenAI-compatible, but `chatMessageSchema.content` was declared `.nullable()` — which allows `null` but still **requires the `content` key to be present**. OpenAI treats assistant `content` as **optional when `tool_calls` is present**, so OpenAI-compatible clients that omit the field entirely (hand-rolled JSON, Go `omitempty`, trimmed payloads) were rejected with `400 INVALID_INPUT` before the request ever reached the provider — even though the identical request with `"content": null` succeeded.

This relaxes `content` to `.nullish()` (nullable **and** optional). It matches OpenAI's contract, stays consistent with the schema's existing permissiveness (it already accepts `content: null` for any role), and `formatMessages` already coerces a missing assistant content to `null`, so no downstream change is required.

Closes #1748

### Changes
- `packages/shared-schemas/src/ai-api.schema.ts` — `chatMessageSchema.content`: `.nullable()` → `.nullish()` (+ explanatory comment).
- `backend/tests/unit/ai-tool-calling-schemas.test.ts` — regression tests (schema-level and request-level).

## How did you test this change?

Added two regression tests to `ai-tool-calling-schemas.test.ts`:

1. `chatMessageSchema` accepts an assistant `tool_calls` message that omits `content` entirely.
2. `chatCompletionRequestSchema` accepts a full multi-turn tool conversation whose assistant turn omits `content`.

Both **fail on `main`** (`Required` at `messages[...].content`) and **pass with the fix**.

Commands run:
- `npx vitest run tests/unit/ai-tool-calling-schemas.test.ts` → **26 passed** (24 existing + 2 new).
- AI gateway suite (`ai-tool-calling-*`, `ai-routes`, `ai-helpers`, `ai-model.service`) → **69 passed**, no regressions; `formatMessages` behavior unchanged.
- `tsc --noEmit` on `packages/shared-schemas` and `backend` → clean (the now-optional `content` type breaks no consumers).
- `eslint` on the changed files → clean.

---
## Summary by cubic
Accept assistant tool-call messages that omit `content`, while keeping `content` required for other roles. Adds role-aware validation to match OpenAI and prevent 400 INVALID_INPUT or silent provider errors.

- **Bug Fixes**
  - In `packages/shared-schemas/src/ai-api.schema.ts`, changed `chatMessageSchema.content` to `.nullish()` and added a role-aware `superRefine` so only `assistant` may omit `content`; `null` stays allowed for all roles.
  - Updated `backend/tests/unit/ai-tool-calling-schemas.test.ts` to cover accepted assistant cases (schema and request) and reject missing `content` on `user`/`tool` messages.

<sup>Written for commit 602cfcf5ed6f2032b3b8538eec62d6f739ead0d2. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/InsForge/InsForge/pull/1749?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

> [!NOTE]
> ### Fix `chatMessageSchema` to accept assistant messages that omit `content`
> Changes the `content` field in `chatMessageSchema` from `.nullable()` to `.nullish()` in [ai-api.schema.ts](https://github.com/InsForge/InsForge/pull/1749/files#diff-b337ceae5cba47edcb003641f623c0fda08c1249dd6b57720f960f0af21fc78c), so that assistant messages with `tool_calls` but no `content` key pass validation. This matches the OpenAI API behavior where `content` may be absent (not just `null`) when tool calls are present.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized d94e685.</sup>
> 
>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved compatibility with OpenAI-compatible tool-calling conversations.
  * Assistant tool-calling messages can now omit `content` (not required when `tool_calls` are present).
  * Validation now correctly handles `content` being `null` or missing across roles, rejecting missing `content` for non-assistant turns.
  * Multi-turn conversations from assistant tool calls to matching tool responses are validated correctly.

### CODE REVIEW

File: packages/shared-schemas/src/ai-api.schema.ts [hunk 1]
```diff
@@ -97,8 +97,13 @@ export const toolCallSchema = z.object({
...
   role: z.enum(['user', 'assistant', 'system', 'tool']),
-  // New format: content can be string or array of content parts (OpenAI-compatible)
-  content: z.union([z.string(), z.array(contentSchema)]).nullable(),
+  // Content can be a string or an array of content parts (OpenAI-compatible).
+  // Nullable AND optional: OpenAI treats assistant `content` as optional when
+  // `tool_calls` is present, so clients following that convention omit the field
+  // entirely. `.nullish()` (string | array | null | undefined) keeps those valid
+  // tool-call messages from being rejected; `formatMessages` already coerces a
+  // missing assistant content to null.
+  content: z.union([z.string(), z.array(contentSchema)]).nullish(),
```
[the remark this answers is not in the document]
Comment (hassaanch23) [reply packages/shared-schemas/src/ai-api.schema.ts]: Good call — fixed in 602cfcf. Rather than a downstream guard, I enforced OpenAI's per-role rule at the schema: a role-aware `superRefine` now only lets **assistant** messages omit `content`; `user`/`system`/`tool` messages still require the key, so a content-less `tool` message is rejected at validation instead of reaching the provider as `undefined`. `null` stays valid for every role. Added tests for the rejected `tool`/`user` cases.

**Review by jwfing — COMMENTED** [review]
## Review: fix(ai): accept assistant tool-call messages that omit content

**Summary:** A tightly-scoped, well-tested schema fix that relaxes `chatMessageSchema.content` to `.nullish()` and adds a role-aware `superRefine` so only `assistant` may omit `content` — correctly matching OpenAI's contract; I found no blocking issues.

**Requirements context:** InsForge/InsForge has no `docs/superpowers/` or `docs/specs/` directory, so no matching spec/plan was found — assessed against the PR description and linked issue #1748. Verified the OpenAI OpenAI-compatible contract claim (assistant `content` optional when `tool_calls` present) against the code paths that consume the schema.

### Findings

#### Critical
(none)

#### Suggestion

**Functionality — `superRefine` is broader than the stated contract (`packages/shared-schemas/src/ai-api.schema.ts:110-125`)**
The refinement only rejects an omitted `content` for non-assistant roles, so it permits an assistant message that omits `content` *unconditionally* — including one that also has no `tool_calls`. OpenAI makes assistant `content` optional **only when `tool_calls` (or `function_call`) is present**. A degenerate `{ role: 'assistant' }` (no content, no tool_calls) now passes validation, then falls through to the non-tool-call branch of `formatMessages` (`backend/src/services/ai/chat-completion.service.ts:133-138`) which sets `content: msg.content` = `undefined` and forwards it to the provider, rather than being caught at the gateway as it was before. Blast radius is low (a malformed message the provider would reject anyway), but tightening the guard would match the PR's own stated intent exactly, e.g.:
```ts
if (
  message.content === undefined &&
  !(message.role === 'assistant' && (message.tool_calls?.length ?? 0) > 0)
) { ctx.addIssue(...) }
```
Note the tool-call branch at `chat-completion.service.ts:112` already coerces omitted content to `null` correctly — this only concerns the assistant-without-tool_calls case.

**Software engineering — test coverage for two edge cases (`backend/tests/unit/ai-tool-calling-schemas.test.ts`)**
The added tests are good (assistant-omits-content accepted; user/tool omission rejected; null still accepted for all roles; full multi-turn request). Two gaps worth covering: (1) the assistant-omits-both-content-and-tool_calls case described above, to pin down the intended contract; (2) an assertion that `formatMessages` coerces an omitted assistant `content` to `null` — the PR relies on this behavior but no test in the diff exercises it directly.

#### Information

- **Stale PR description.** The top "Summary" section describes a plain `.nullable()` → `.nullish()` change, but the shipped commit (`602cfcf`) also adds the role-aware `superRefine`. The cubic and CodeRabbit summaries are accurate; consider updating the hand-written summary so reviewers/history reflect the actual change.
- **`ZodObject` → `ZodEffects` is safe here.** Adding `.superRefine` changes the schema's runtime type, which would break `.extend`/`.merge`/`.pick`/`.shape` consumers — I confirmed there are none; `chatMessageSchema` is only array-wrapped in `chatCompletionRequestSchema` and consumed via `.safeParse`, so nesting/validation is unaffected.
- **Security:** No security-relevant changes — this is a validation-loosening on an existing field with no new sink (SQL/shell/HTTP), no secrets/PII, and no auth changes; the `superRefine` keeps non-assistant roles strict.
- **Performance:** No concern — the refinement is an O(1) per-message check with no added queries, loops, or allocations.

### Verdict

**approved** (informational — no Critical findings; the two Suggestions are non-blocking). The core fix is correct, matches the OpenAI-compatible contract for the intended tool-calling case, and is backed by meaningful regression tests. Human approval via the GitHub approve flow is still a separate step.

**Review by jwfing — APPROVED** [review]
